### PR TITLE
Photoshop: resize saved images in ExtractReview for ffmpeg

### PR DESCRIPTION
**Pull request by kalisp**
## Brief description
Ffmpeg cannot handle pictures higher than 16384x16384. 

## Description
Used PIL to resize to max size(with aspect ratio).

## Additional info
Implemented as a 'hotfix' to unblock customers production.
In the future this plugin should be refactored (to use general ExtractThumbnail and ExtractReview). That is currently blocked by not having oiio on all platforms.

## Testing notes:
1. publish workfile with larger resolution than 16384x16384 in PS

### CODE REVIEW

File: openpype/hosts/photoshop/plugins/publish/extract_review.py [hunk 1]
```diff
@@ -49,34 +57,41 @@ def process(self, instance):
...
+            processed_img_names = [img_list]
 
         ffmpeg_path = openpype.lib.get_ffmpeg_tool_path("ffmpeg")
 
         instance.data["stagingDir"] = staging_dir
 
-        # Generate thumbnail.
+        source_files_pattern = os.path.join(staging_dir,
+                                            self.output_seq_filename)
+        source_files_pattern = self._check_and_resize(processed_img_names,
```
Comment (iLLiCiTiT) [inline openpype/hosts/photoshop/plugins/publish/extract_review.py:80]: Resized temp files won't be cleaned up. They should be added to `context.data["cleanupFullPaths"]` or removed here when are not needed anymore.

**Review by m-u-r-p-h-y — CHANGES_REQUESTED** [review]
```

Instance: 
review
Message: 
Image size (441000000 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.
Line: 
3006
Traceback: 
Traceback (most recent call last):
  File "D:\REPO\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 522, in __explicit_process
    runner(*args)
  File "D:\REPO\OpenPype\openpype\hosts\photoshop\plugins\publish\extract_review.py", line 82, in process
  File "D:\REPO\OpenPype\openpype\hosts\photoshop\plugins\publish\extract_review.py", line 148, in _check_and_resize
  File "D:\REPO\OpenPype\.venv\lib\site-packages\PIL\Image.py", line 3109, in open
    im = _open_core(fp, filename, prefix, formats)
  File "D:\REPO\OpenPype\.venv\lib\site-packages\PIL\Image.py", line 3096, in _open_core
    _decompression_bomb_check(im.size)
  File "D:\REPO\OpenPype\.venv\lib\site-packages\PIL\Image.py", line 3006, in _decompression_bomb_check
    f"Image size ({pixels} pixels) exceeds limit of {2 * MAX_IMAGE_PIXELS} "
PIL.Image.DecompressionBombError: Image size (441000000 pixels) exceeds limit of 178956970 pixels, could be decompression bomb DOS attack.

```



image

**Review by m-u-r-p-h-y — COMMENTED** [review]
there is still an issue with ffmpeg conversion. this time it is memory allocation error


```
Input #0, image2, from 'C:\Users\murph\AppData\Local\Temp\pyblish_tmp_0rrgjly2\resized\22101_sh010_lighting_v002.%04d.jpg': 
  Duration: 00:00:00.04, start: 0.000000, bitrate: N/A 
  Stream #0:0: Video: mjpeg (Baseline), yuvj420p(pc, bt470bg/unknown/unknown), 16000x16000 [SAR 1:1 DAR 1:1], 25 fps, 25 tbr, 25 tbn, 25 tbc 
Stream mapping: 
  Stream #0:0 -> #0:0 (mjpeg (native) -> h264 (libx264)) 
Press [q] to stop, [?] for help 
[libx264 @ 0000025b24f75f80] using SAR=1/1 
[libx264 @ 0000025b24f75f80] frame MB size (1000x1000) > level limit (139264) 
[libx264 @ 0000025b24f75f80] DPB size (4 frames, 4000000 mbs) > level limit (0 frames, 696320 mbs) 
[libx264 @ 0000025b24f75f80] MB rate (25000000) > level limit (16711680) 
[libx264 @ 0000025b24f75f80] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2 
x264 [error]: malloc of size 440626798 failed 
Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height 
Conversion failed! 
```

File: openpype/hosts/photoshop/plugins/publish/extract_review.py [hunk 2]
```diff
@@ -120,6 +136,35 @@ def process(self, instance):
...
 
+    def _check_and_resize(self, processed_img_names, source_files_pattern,
+                          staging_dir):
+        """Check if saved image could be used in ffmpeg.
+
+        Ffmpeg has max size 16384x16384. Saved image(s) must be resized to be
+        used as a source for thumbnail or review mov.
+        """
+        # 16384x16384 actually didn't work because int overflow
+        max_ffmpeg_size = 16000
```
Comment (iLLiCiTiT) [inline openpype/hosts/photoshop/plugins/publish/extract_review.py:147]: Maybe let's use 8K (8192).

**Comment by m-u-r-p-h-y** [comment]
it is weird anyway 

`x264 [error]: malloc of size 440626798 failed `

what is the problem with 440MB?

**Comment by iLLiCiTiT** [comment]
That happens when you run out of RAM.

**Comment by m-u-r-p-h-y** [comment]
I still have over 14GB ram available (out of 32)

screengrabbed just after the failure
image

File: openpype/hosts/photoshop/plugins/publish/extract_review.py (hunk 1 above)
Comment (kalisp) [reply openpype/hosts/photoshop/plugins/publish/extract_review.py:80]: It is in "stagingDir", so they will be removed by "cleanup.py", I checked.

File: openpype/hosts/photoshop/plugins/publish/extract_review.py (hunk 2 above)
Comment (kalisp) [reply openpype/hosts/photoshop/plugins/publish/extract_review.py]: changed to 8k. It actually worked with 16K on our testing PC, but I guess 8K is plenty for review purposes.

**Comment by BigRoy** [comment]
> That happens when you run out of RAM.

Technically, that would happen when the application is unable to allocate that RAM. You might have free ram whilst the application cannot acquire it. (I have limited knowledge on this topic, but I believe this is the case).

Anyway, [the issue you're facing sounds like this one on StackOverflow](https://stackoverflow.com/a/22572790).

**Comment by m-u-r-p-h-y** [comment]
Hmm it would mean x264 codec will have troubles even with 4k

Interestingly Peter @kalisp was able to use 16k. How come?

**Comment by BigRoy** [comment]
> Interestingly Peter @kalisp was able to use 16k. How come?

The issue might have been solely with x264 codec in older FFMPEG versions of course - the encoder could've been updated in the meantime. The error message felt suspiciously close to what was reported there however.

**Comment by m-u-r-p-h-y** [comment]
Shouldn't we update our vendorized ffmpeg binary in OP? Our version is 4.4 compared to latest 5.1

**Comment by m-u-r-p-h-y** [comment]
and also if the max resolution of review is behaving so unpredictable, we may want to put this into settings so there is an easy way of changing it, depending on the hw and sw limitations on client side. (plus different OS influence)

**Review by m-u-r-p-h-y — COMMENTED** [review]
It works on my machine with 8k limit

with both versions of ffmpeg 4.4 and 5.1

**Comment by kalisp** [comment]
Added to Settings:
Screenshot 2022-08-22 111439

**Review by m-u-r-p-h-y — APPROVED** [review]
The settings are applied 

image

**Pull request merged**